### PR TITLE
🚀 HTTP/2 SUPPORT

### PR DESCRIPTION
**Pull request by balcieren**
Usage :
```go
	app := fiber.New(fiber.Config{
		HTTP2: true,
                HTTP2Config: fiber.HTTP2Config{...},
	})

	app.Get("/http11", func(c *fiber.Ctx) error {
		return c.JSON(c.IsHTTP11()) //return true without http2 client support
	})

	app.Get("/http2", func(c *fiber.Ctx) error {
		return c.JSON(c.IsHTTP2()) //return true via http2 client support
	})

	app.ListenTLS(":8080", ".github/testdata/public.crt", ".github/testdata/private.key")
```

Issue #1683

### To Do List

- [x] Add property to app's config.
- [x] ListenTLS with HTTP2.
- [x] Add IsHTTP2 and IsHTTP11 functions to context.
- [ ] HTTP2 client. (fiber client is still in beta)
- [x] Add tests.

ZAAAA

### CODE REVIEW

File: ctx.go [hunk 1]
```diff
@@ -1343,3 +1349,13 @@ func (c *Ctx) IsFromLocal() bool {
 	}
 	return c.isLocalHost(ips[0])
 }
+
+//IsRequestHTTP2 will return true if request's header protocol equals to 'HTTP/2'
+func (c *Ctx) IsRequestHTTP2() bool {
+	return string(c.Request().Header.Protocol()) == HTTP2
```
Comment (ReneWerner87) [inline ctx.go:1355]: please use https://github.com/gofiber/fiber/blob/master/utils/convert.go#L16

File: ctx.go [hunk 2]
```diff
@@ -1343,3 +1349,13 @@ func (c *Ctx) IsFromLocal() bool {
...
+}
+
+//IsRequestHTTP11 will return true if request's header protocol equals to 'HTTP/1.1'
+func (c *Ctx) IsRequestHTTP11() bool {
+	return string(c.Request().Header.Protocol()) == HTTP11
```
Comment (ReneWerner87) [inline ctx.go:1360]: please use https://github.com/gofiber/fiber/blob/master/utils/convert.go#L16

**Review by renanbastos93 — APPROVED** [review]
Wow, good job.
image

**Comment by balcieren** [comment]
> @balcieren could we rename `EnableHTTP2` to `HTTP2`? Same goes for context method, `IsHTTP2` / `IsHTTP1` (Or `c.Protocol()` @ReneWerner87 thoughts?)

I am okay if you and @ReneWerner87 are okay. I can commit tonight. Also c.Protocol is returning "http" or "https" but fasthttp's protocol is returning "HTTP/1.1" or "HTTP/2.0". We should decide it.

**Comment by balcieren** [comment]
> @balcieren could we rename `EnableHTTP2` to `HTTP2`? Same goes for context method, `IsHTTP2` / `IsHTTP1` (Or `c.Protocol()` @ReneWerner87 thoughts?)

Also I agree to @efectn . Firstly I have made IsHTTP2 but after that I have seen in fasthttp's doc, Request header and reponse header can return Proto. So I changed it with IsRequestHTTP2.

**Comment by dgrr** [comment]
I'd wait a little bit. When I said it's `production ready` I was overrating my library tbh. It's not production ready if you are serving content to users and security is key. For example, I wouldn't put this library in front of an API for a crypto exchange.
Truth is that I am currently using the client in production environment, yes, but my own implementation of the client using the http2 library.

The library might be poorly organized. This library aims to be easy for users who just want to setup http/2 clients/servers like, for instance, a GRPC library, and users who just want to provide HTTP/2 on their fasthttp programs. Because of that, you can see in the doc that almost every function is accessible from the user perspective.

Now, what the roadmap is? Last year I've been very busy, and working in something else other than my job wasn't possible. Even reading and replying to issues wasn't a thing. Now that I have more time I've decided to start taking the library to the next level. 
Thanks to some help from @juliens we have the testing tools to evaluate the capacity of the library on being production ready. 
First goal is to pass most of the tests that are in-place already. Some tests are not suitable for the library because they can't be completed, and some cases are not present in the tests, like https://github.com/dgrr/http2/issues/52.

Once the tests are ready, we can start thinking about production. I'll try to add debugging information everywhere so users can debug and report issues to the [main repo](https://github.com/dgrr/http2). For now, and I think for the future, [fasthttp's http2 repo](https://github.com/fasthttp/http2) is going to be a mirror of the main repository.

**Comment by balcieren** [comment]
> I'd wait a little bit. When I said it's `production ready` I was overrating my library tbh. It's not production ready if you are serving content to users and security is key. For example, I wouldn't put this library in front of an API for a crypto exchange. Truth is that I am currently using the client in production environment, yes, but my own implementation of the client using the http2 library.
> 
> The library might be poorly organized. This library aims to be easy for users who just want to setup http/2 clients/servers like, for instance, a GRPC library, and users who just want to provide HTTP/2 on their fasthttp programs. Because of that, you can see in the doc that almost every function is accessible from the user perspective.
> 
> Now, what the roadmap is? Last year I've been very busy, and working in something else other than my job wasn't possible. Even reading and replying to issues wasn't a thing. Now that I have more time I've decided to start taking the library to the next level. Thanks to some help from @juliens we have the testing tools to evaluate the capacity of the library on being production ready. First goal is to pass most of the tests that are in-place already. Some tests are not suitable for the library because they can't be completed, and some cases are not present in the tests, like [dgrr/http2#52](https://github.com/dgrr/http2/issues/52).
> 
> Once the tests are ready, we can start thinking about production. I'll try to add debugging information everywhere so users can debug and report issues to the [main repo](https://github.com/dgrr/http2). For now, and I think for the future, [fasthttp's http2 repo](https://github.com/fasthttp/http2) is going to be a mirror of the main repository.

I agree you. We should wait stable version.

**Comment by efectn** [comment]
@ReneWerner87 @Fenny what do you think about that? I think we can convert this pr draft, until dgrr/http2's server side is stable.

**Comment by ReneWerner87** [comment]
Yes with pleasure, we can convert it to draft

**Comment by xdorro** [comment]
I'm waiting for this feature

**Comment by balcieren** [comment]
moved https://github.com/gofiber/fiber/pull/1796

**Pull request closed without merging**